File: app/models/patient.rb
class Patient < ApplicationRecord
  has_many :appointments, dependent: :destroy
  has_many :doctors, through: :appointments

  # def get_doctors_with_appointments
  #   a = self.doctors.map do |doctor|
  #    appointment = doctor.appointments.find_by(patient_id: self.id)
  #     {doctor: doctor.name, appDate: appointment.date, appId: appointment.id}
  #   end
  #   return {patient: self.name, appointment: a}
  # end

  def get_doctors_with_appointments
    a = self.appointments.map do |app|
      # appointment = app.doctors.find_by(patient_id: self.id)
      {doctor: app.doctor.name, appDate: app.date, id: app.id}
    end
    return {patient: self.name, appointment: a}
  end


end

File: client/src/pages/patients/Patient.js
import axios from 'axios'
import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import ErrorMessage from '../../components/ErrorMessage'
import List from '../../components/List'
import Spinner from '../../components/Spinner'
import useAxiosOnMount from '../../customHooks/useAxiosOnMount'
import PatientForm from './PatientForm'

const Patient = () => {
  const {id} = useParams()
  const {data, loading, error, setData} = useAxiosOnMount(`/api/patients/${id}`)
  const [doctors, setDoctors] = useState([])

  useEffect(()=>{
    getDoctors()
  }, [])

  
    const getDoctors = async () => {
      let res = await axios.get(`/api/doctors`)
      setDoctors(res.data)
    }

    const formatDoctorsData = () => {
      if (!doctors) return []
      return doctors.map(d =>{
          return {key:d.id, text:d.name, value:d.id}
      })
    }

    const addApp = (app) => {
      console.log(app)
      console.log('data',data)
      setData([...data, app])
    }
    
    if(loading) return <Spinner />
    if(error) return <ErrorMessage error={error}/>
    return(
    <div>
      <div>
        <PatientForm
          addApp={addApp}
          patientId={id} 
          patientName={data.patient} 
          doctors={doctors} 
          doctorsData={formatDoctorsData()}
        />
      </div>
      <List 
        name={`Appointments for ${data.patient}`}
        data={data.appointment}
        renderData = {(a)=> {
          return (
            <div key={a.id} style={{marginBottom: '15px'}}>
              <h2>{a.doctor}</h2>
              <h4>{a.appDate}</h4>
           </div>
          )
        }}
      />
    </div>
  )
}

export default Patient